src/library/scala/collection/immutable/ArraySeq.scala

```diff
@@ -77,12 +77,65 @@ sealed abstract class ArraySeq[+A]
     ArraySeq.unsafeWrapArray(unsafeArray.appended[Any](elem)).asInstanceOf[ArraySeq[B]]
 
   override def appendedAll[B >: A](suffix: collection.IterableOnce[B]): ArraySeq[B] = {
-    val b = ArrayBuilder.make[Any]
-    val k = suffix.knownSize
-    if(k >= 0) b.sizeHint(k + unsafeArray.length)
-    b.addAll(unsafeArray)
-    b.addAll(suffix)
-    ArraySeq.unsafeWrapArray(b.result()).asInstanceOf[ArraySeq[B]]
+    def genericResult(it: Iterator[B]) = {
+      val b = ArrayBuilder.make[Any]
+      val k = suffix.knownSize
+      if(k >= 0) b.sizeHint(k + unsafeArray.length)
+      b.addAll(unsafeArray)
+      b.addAll(it)
```

Comment:
I fear that passing the iterator here will degrade performance in some cases. For example, if you append a mutable ArraySeq

https://github.com/scala/scala/blob/dfb8f7364a56fc01ad1eaea57980fba586ef6914/src/library/scala/collection/mutable/ArrayBuilder.scala#L60-L65

ends up at

https://github.com/scala/scala/blob/dfb8f7364a56fc01ad1eaea57980fba586ef6914/src/library/scala/collection/mutable/ArraySeq.scala#L79-L82

Reply:
Great catch! I'll fix that.

Reply:
Updated :)